File: interpreter/parser.go
package interpreter

import (
	"fmt"
	"os"
	"strconv"
)

type Parser struct {
	tokens []Token
	pos    int
}

func NewParser(tokens []Token) *Parser {
	return &Parser{tokens: tokens}
}

func (p *Parser) peek() Token {
	if p.pos >= len(p.tokens) {
		return Token{TOKEN_EOF, "", 0, 0}
	}
	return p.tokens[p.pos]
}

func (p *Parser) advance() Token {
	t := p.tokens[p.pos]
	p.pos++
	return t
}

func (p *Parser) expect(tt TokenType) Token {
	t := p.peek()
	if t.Type != tt {
		fmt.Fprintf(os.Stderr, "line %d:%d: expected %s, got %s (%q)\n",
			t.Line, t.Col, tokenNames[tt], tokenNames[t.Type], t.Value)
		os.Exit(1)
	}
	return p.advance()
}

func (p *Parser) match(tt TokenType) bool {
	if p.peek().Type == tt {
		p.advance()
		return true
	}
	return false
}

func (p *Parser) atIdent(v string) bool {
	return p.peek().Type == TOKEN_IDENT && p.peek().Value == v
}

func (p *Parser) parseType() BType {
	t := p.expect(TOKEN_IDENT)
	switch t.Value {
	case "float":
		return TYPE_FLOAT
	case "string":
		return TYPE_STRING
	case "bool":
		return TYPE_BOOL
	case "bel":
		return TYPE_BEL
	case "ster":
		return TYPE_STER
	case "any":
		return TYPE_ANY
	default:
		fmt.Fprintf(os.Stderr, "line %d:%d: unknown type '%s'\n", t.Line, t.Col, t.Value)
		os.Exit(1)
	}
	return TYPE_ANY
}

func (p *Parser) Parse() *Program {
	prog := &Program{}
	for p.peek().Type != TOKEN_EOF {
		prog.Stmts = append(prog.Stmts, p.parseStmt())
	}
	return prog
}

func (p *Parser) parseStmt() ASTNode {
	if p.atIdent("var") {
		return p.parseVarDecl()
	}
	if p.atIdent("fn") {
		return p.parseFnDecl()
	}
	if p.atIdent("if") {
		return p.parseIf()
	}
	if p.atIdent("while") {
		return p.parseWhile()
	}
	if p.atIdent("for") {
		return p.parseFor()
	}
	if p.atIdent("return") {
		return p.parseReturn()
	}
	if p.atIdent("break") {
		p.advance()
		p.match(TOKEN_SEMICOLON)
		return &BreakStmt{}
	}
	if p.atIdent("continue") {
		p.advance()
		p.match(TOKEN_SEMICOLON)
		return &ContinueStmt{}
	}
	if p.peek().Type == TOKEN_LBRACE {
		return p.parseBlock()
	}

	if p.atIdent("idb") {
		return p.parseIdbDecl()
	}

	if p.peek().Type == TOKEN_IDENT && isTypeName(p.peek().Value) {
		return p.parseTypedDecl()
	}

	expr := p.parseExpr()

	if p.peek().Type == TOKEN_EQ {
		p.advance()
		val := p.parseExpr()
		p.match(TOKEN_SEMICOLON)
		return &Assign{Target: expr, Value: val}
	}
	if p.peek().Type == TOKEN_PLUS_EQ {
		p.advance()
		val := p.parseExpr()
		p.match(TOKEN_SEMICOLON)
		return &Assign{Target: expr, Value: &BinExpr{Op: "+", Left: expr, Right: val}}
	}
	if p.peek().Type == TOKEN_MINUS_EQ {
		p.advance()
		val := p.parseExpr()
		p.match(TOKEN_SEMICOLON)
		return &Assign{Target: expr, Value: &BinExpr{Op: "-", Left: expr, Right: val}}
	}

	p.match(TOKEN_SEMICOLON)
	return &ExprStmt{Expr: expr}
}

func (p *Parser) parseBlock() ASTNode {
	p.expect(TOKEN_LBRACE)
	b := &BlockStmt{}
	for p.peek().Type != TOKEN_RBRACE && p.peek().Type != TOKEN_EOF {
		b.Body = append(b.Body, p.parseStmt())
	}
	p.expect(TOKEN_RBRACE)
	return b
}

func (p *Parser) parseIf() ASTNode {
	p.advance()
	cond := p.parseExpr()
	body := p.parseBlock()
	var els ASTNode
	if p.atIdent("else") {
		p.advance()
		if p.atIdent("if") {
			els = p.parseIf()
		} else {
			els = p.parseBlock()
		}
	}
	return &IfStmt{Cond: cond, Body: body, Else: els}
}

func (p *Parser) parseWhile() ASTNode {
	p.advance()
	cond := p.parseExpr()
	body := p.parseBlock()
	return &WhileStmt{Cond: cond, Body: body}
}

func (p *Parser) parseFor() ASTNode {
	p.advance()
	varName := p.expect(TOKEN_IDENT).Value
	p.expect(TOKEN_IDENT) // "in"
	iter := p.parseExpr()
	body := p.parseBlock()
	return &ForStmt{Var: varName, Iter: iter, Body: body}
}

func (p *Parser) parseVarDecl() ASTNode {
	p.advance()
	name := p.expect(TOKEN_IDENT).Value

	var varType BType
	hasType := false
	if p.match(TOKEN_COLON) {
		varType = p.parseType()
		hasType = true
	}

	p.expect(TOKEN_EQ)
	val := p.parseExpr()
	p.match(TOKEN_SEMICOLON)
	return &VarDecl{Name: name, VarType: varType, Value: val, HasType: hasType}
}

func (p *Parser) parseIdbDecl() ASTNode {
	p.advance()
	num := p.expect(TOKEN_NUMBER)
	id, err := strconv.Atoi(num.Value)
	if err != nil {
		fmt.Fprintf(os.Stderr, "line %d:%d: invalid idb id '%s'\n", num.Line, num.Col, num.Value)
		os.Exit(1)
	}
	p.expect(TOKEN_EQ)
	val := p.parseExpr()
	p.match(TOKEN_SEMICOLON)
	return &IdbDecl{Id: id, Value: val}
}

func isTypeName(name string) bool {
	switch name {
	case "float", "string", "bool", "bel", "ster", "any":
		return true
	}
	return false
}

func (p *Parser) parseTypedDecl() ASTNode {
	typeName := p.advance().Value
	name := p.expect(TOKEN_IDENT).Value
	p.expect(TOKEN_EQ)
	val := p.parseExpr()
	p.match(TOKEN_SEMICOLON)

	var btype BType
	isBel := false
	isSter := false
	switch typeName {
	case "float":
		btype = TYPE_FLOAT
	case "string":
		btype = TYPE_STRING
	case "bool":
		btype = TYPE_BOOL
	case "bel":
		btype = TYPE_BEL
		isBel = true
	case "ster":
		btype = TYPE_STER
		isSter = true
	case "any":
		btype = TYPE_ANY
	}
	return &VarDecl{Name: name, VarType: btype, Value: val, HasType: true, IsBel: isBel, IsSter: isSter}
}

func (p *Parser) parseFnDecl() ASTNode {
	p.advance()
	name := p.expect(TOKEN_IDENT).Value
	p.expect(TOKEN_LPAREN)

	var params []TypedParam
	if p.peek().Type != TOKEN_RPAREN {
		pName := p.expect(TOKEN_IDENT).Value
		pType := TYPE_ANY
		if p.match(TOKEN_COLON) {
			pType = p.parseType()
		}
		params = append(params, TypedParam{Name: pName, Type: pType})
		for p.match(TOKEN_COMMA) {
			pName = p.expect(TOKEN_IDENT).Value
			pType = TYPE_ANY
			if p.match(TOKEN_COLON) {
				pType = p.parseType()
			}
			params = append(params, TypedParam{Name: pName, Type: pType})
		}
	}
	p.expect(TOKEN_RPAREN)

	var retType BType
	hasRetType := false
	if p.match(TOKEN_COLON) {
		retType = p.parseType()
		hasRetType = true
	}

	body := p.parseBlock()
	return &FnDecl{Name: name, Params: params, ReturnType: retType, HasRetType: hasRetType, Body: body}
}

func (p *Parser) parseReturn() ASTNode {
	p.advance()
	var val ASTNode
	if p.peek().Type != TOKEN_SEMICOLON && p.peek().Type != TOKEN_RBRACE && p.peek().Type != TOKEN_EOF {
		val = p.parseExpr()
	}
	p.match(TOKEN_SEMICOLON)
	return &ReturnStmt{Value: val}
}

func (p *Parser) parseExpr() ASTNode { return p.parseOr() }

func (p *Parser) parseOr() ASTNode {
	left := p.parseAnd()
	for p.peek().Type == TOKEN_OR {
		p.advance()
		right := p.parseAnd()
		left = &BinExpr{Op: "||", Left: left, Right: right}
	}
	return left
}

func (p *Parser) parseAnd() ASTNode {
	left := p.parseComparison()
	for p.peek().Type == TOKEN_AND {
		p.advance()
		right := p.parseComparison()
		left = &BinExpr{Op: "&&", Left: left, Right: right}
	}
	return left
}

func (p *Parser) parseComparison() ASTNode {
	left := p.parseAddSub()
	for {
		t := p.peek()
		if t.Type == TOKEN_EQEQ || t.Type == TOKEN_NEQ ||
			t.Type == TOKEN_LT || t.Type == TOKEN_GT ||
			t.Type == TOKEN_LTE || t.Type == TOKEN_GTE {
			p.advance()
			right := p.parseAddSub()
			left = &BinExpr{Op: t.Value, Left: left, Right: right}
		} else {
			break
		}
	}
	return left
}

func (p *Parser) parseAddSub() ASTNode {
	left := p.parseMulDiv()
	for {
		t := p.peek()
		if t.Type == TOKEN_PLUS || t.Type == TOKEN_MINUS {
			p.advance()
			right := p.parseMulDiv()
			left = &BinExpr{Op: t.Value, Left: left, Right: right}
		} else {
			break
		}
	}
	return left
}

func (p *Parser) parseMulDiv() ASTNode {
	left := p.parseUnary()
	for {
		t := p.peek()
		if t.Type == TOKEN_STAR || t.Type == TOKEN_SLASH || t.Type == TOKEN_PERCENT {
			p.advance()
			right := p.parseUnary()
			left = &BinExpr{Op: t.Value, Left: left, Right: right}
		} else {
			break
		}
	}
	return left
}

func (p *Parser) parseUnary() ASTNode {
	if p.peek().Type == TOKEN_MINUS {
		p.advance()
		return &UnaryExpr{Op: "-", Expr: p.parsePostfix()}
	}
	if p.peek().Type == TOKEN_NOT {
		p.advance()
		return &UnaryExpr{Op: "!", Expr: p.parsePostfix()}
	}
	return p.parsePostfix()
}

func (p *Parser) parsePostfix() ASTNode {
	node := p.parsePrimary()
	for {
		if p.peek().Type == TOKEN_LPAREN {
			p.advance()
			var call *CallExpr
			switch n := node.(type) {
			case *CallExpr:
				call = n
			case *Ident:
				call = &CallExpr{Name: n.Name}
			default:
				fmt.Fprintf(os.Stderr, "line %d: not callable\n", p.peek().Line)
				os.Exit(1)
			}
			if p.peek().Type != TOKEN_RPAREN {
				call.Args = append(call.Args, p.parseExpr())
				for p.match(TOKEN_COMMA) {
					call.Args = append(call.Args, p.parseExpr())
				}
			}
			p.expect(TOKEN_RPAREN)
			node = call
		} else if p.peek().Type == TOKEN_DOT {
			p.advance()
			right := p.expect(TOKEN_IDENT).Value
			node = &DotExpr{Left: node, Right: right}
		} else if p.peek().Type == TOKEN_LBRACKET {
			p.advance()
			idx := p.parseExpr()
			p.expect(TOKEN_RBRACKET)
			node = &IndexExpr{Obj: node, Index: idx}
		} else {
			break
		}
	}
	return node
}

func (p *Parser) parsePrimary() ASTNode {
	t := p.peek()
	switch t.Type {
	case TOKEN_STRING:
		p.advance()
		return &StringLit{Value: t.Value}
	case TOKEN_NUMBER:
		p.advance()
		return &NumLit{Value: t.Value}
	case TOKEN_IDENT:
		p.advance()
		switch t.Value {
		case "true":
			return &BoolLit{Value: true}
		case "false":
			return &BoolLit{Value: false}
		case "null":
			return &Ident{Name: "null"}
		default:
			return &Ident{Name: t.Value}
		}
	case TOKEN_LPAREN:
		p.advance()
		expr := p.parseExpr()
		p.expect(TOKEN_RPAREN)
		return expr
	case TOKEN_LBRACKET:
		return p.parseArrayLit()
	default:
		fmt.Fprintf(os.Stderr, "line %d:%d: unexpected token '%s' (%q)\n",
			t.Line, t.Col, tokenNames[t.Type], t.Value)
		os.Exit(1)
	}
	return nil
}

func (p *Parser) parseArrayLit() ASTNode {
	p.expect(TOKEN_LBRACKET)
	al := &ArrayLit{}
	if p.peek().Type != TOKEN_RBRACKET {
		al.Elems = append(al.Elems, p.parseExpr())
		for p.match(TOKEN_COMMA) {
			al.Elems = append(al.Elems, p.parseExpr())
		}
	}
	p.expect(TOKEN_RBRACKET)
	return al
}

File: src/parser.rs
use crate::ast::*;
use crate::token::{Token, TokenType};
use crate::types::BType;

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

fn is_type_name(name: &str) -> bool {
    matches!(name, "float" | "string" | "bool" | "bel" | "ster" | "any")
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, pos: 0 }
    }

    fn peek(&self) -> &Token {
        if self.pos >= self.tokens.len() {
            static EOF: Token = Token {
                tt: TokenType::Eof,
                value: String::new(),
                line: 0,
                col: 0,
            };
            &EOF
        } else {
            &self.tokens[self.pos]
        }
    }

    fn advance(&mut self) -> Token {
        let t = self.tokens[self.pos].clone();
        self.pos += 1;
        t
    }

    fn expect(&mut self, tt: TokenType) -> Token {
        let t = self.peek().clone();
        if t.tt != tt {
            panic!(
                "line {}:{}: expected {}, got {} ({})",
                t.line,
                t.col,
                tt.name(),
                t.tt.name(),
                t.value
            );
        }
        self.advance()
    }

    fn match_token(&mut self, tt: TokenType) -> bool {
        if self.peek().tt == tt {
            self.advance();
            true
        } else {
            false
        }
    }

    fn at_ident(&self, v: &str) -> bool {
        self.peek().tt == TokenType::Ident && self.peek().value == v
    }

    fn parse_type(&mut self) -> BType {
        let t = self.expect(TokenType::Ident);
        match t.value.as_str() {
            "float" => BType::Float,
            "string" => BType::String,
            "bool" => BType::Bool,
            "bel" => BType::Bel,
            "ster" => BType::Ster,
            "any" => BType::Any,
            _ => {
                panic!("line {}:{}: unknown type '{}'", t.line, t.col, t.value);
            }
        }
    }

    pub fn parse(&mut self) -> Program {
        let mut prog = Program { stmts: Vec::new() };
        while self.peek().tt != TokenType::Eof {
            prog.stmts.push(self.parse_stmt());
        }
        prog
    }

    fn parse_stmt(&mut self) -> ASTNode {
        if self.at_ident("var") {
            return self.parse_var_decl();
        }
        if self.at_ident("fn") {
            return self.parse_fn_decl();
        }
        if self.at_ident("if") {
            return self.parse_if();
        }
        if self.at_ident("while") {
            return self.parse_while();
        }
        if self.at_ident("for") {
            return self.parse_for();
        }
        if self.at_ident("return") {
            return self.parse_return();
        }
        if self.at_ident("break") {
            self.advance();
            self.match_token(TokenType::Semicolon);
            return ASTNode::BreakStmt;
        }
        if self.at_ident("continue") {
            self.advance();
            self.match_token(TokenType::Semicolon);
            return ASTNode::ContinueStmt;
        }
        if self.peek().tt == TokenType::LBrace {
            return self.parse_block();
        }
        if self.at_ident("idb") {
            return self.parse_idb_decl();
        }
        if self.peek().tt == TokenType::Ident && is_type_name(&self.peek().value) {
            return self.parse_typed_decl();
        }

        let expr = self.parse_expr();

        if self.peek().tt == TokenType::Eq {
            self.advance();
            let val = self.parse_expr();
            self.match_token(TokenType::Semicolon);
            return ASTNode::Assign(Box::new(Assign {
                target: Box::new(expr),
                value: Box::new(val),
            }));
        }
        if self.peek().tt == TokenType::PlusEq {
            self.advance();
            let val = self.parse_expr();
            self.match_token(TokenType::Semicolon);
            return ASTNode::Assign(Box::new(Assign {
                target: Box::new(expr.clone()),
                value: Box::new(ASTNode::BinExpr(Box::new(BinExpr {
                    op: "+".to_string(),
                    left: Box::new(expr),
                    right: Box::new(val),
                }))),
            }));
        }
        if self.peek().tt == TokenType::MinusEq {
            self.advance();
            let val = self.parse_expr();
            self.match_token(TokenType::Semicolon);
            return ASTNode::Assign(Box::new(Assign {
                target: Box::new(expr.clone()),
                value: Box::new(ASTNode::BinExpr(Box::new(BinExpr {
                    op: "-".to_string(),
                    left: Box::new(expr),
                    right: Box::new(val),
                }))),
            }));
        }

        self.match_token(TokenType::Semicolon);
        ASTNode::ExprStmt(Box::new(ExprStmt {
            expr: Box::new(expr),
        }))
    }

    fn parse_block(&mut self) -> ASTNode {
        self.expect(TokenType::LBrace);
        let mut body = Vec::new();
        while self.peek().tt != TokenType::RBrace && self.peek().tt != TokenType::Eof {
            body.push(self.parse_stmt());
        }
        self.expect(TokenType::RBrace);
        ASTNode::BlockStmt(Box::new(BlockStmt { body }))
    }

    fn parse_if(&mut self) -> ASTNode {
        self.advance();
        let cond = self.parse_expr();
        let body = self.parse_block();
        let mut else_body = None;
        if self.at_ident("else") {
            self.advance();
            if self.at_ident("if") {
                else_body = Some(Box::new(self.parse_if()));
            } else {
                else_body = Some(Box::new(self.parse_block()));
            }
        }
        ASTNode::IfStmt(Box::new(IfStmt {
            cond: Box::new(cond),
            body: Box::new(body),
            else_body,
        }))
    }

    fn parse_while(&mut self) -> ASTNode {
        self.advance();
        let cond = self.parse_expr();
        let body = self.parse_block();
        ASTNode::WhileStmt(Box::new(WhileStmt {
            cond: Box::new(cond),
            body: Box::new(body),
        }))
    }

    fn parse_for(&mut self) -> ASTNode {
        self.advance();
        let var_name = self.expect(TokenType::Ident).value;
        self.expect(TokenType::Ident); // "in"
        let iter = self.parse_expr();
        let body = self.parse_block();
        ASTNode::ForStmt(Box::new(ForStmt {
            var: var_name,
            iter: Box::new(iter),
            body: Box::new(body),
        }))
    }

    fn parse_var_decl(&mut self) -> ASTNode {
        self.advance();
        let name = self.expect(TokenType::Ident).value;

        let mut var_type = BType::Any;
        let mut has_type = false;
        if self.match_token(TokenType::Colon) {
            var_type = self.parse_type();
            has_type = true;
        }

        self.expect(TokenType::Eq);
        let val = self.parse_expr();
        self.match_token(TokenType::Semicolon);
        ASTNode::VarDecl(Box::new(VarDecl {
            name,
            var_type,
            value: Box::new(val),
            has_type,
            is_bel: false,
            is_ster: false,
        }))
    }

    fn parse_idb_decl(&mut self) -> ASTNode {
        self.advance();
        let num = self.expect(TokenType::Number);
        let id = num.value.parse::<i64>().unwrap_or(0);
        self.expect(TokenType::Eq);
        let val = self.parse_expr();
        self.match_token(TokenType::Semicolon);
        ASTNode::IdbDecl(Box::new(IdbDecl {
            id,
            value: Box::new(val),
        }))
    }

    fn parse_typed_decl(&mut self) -> ASTNode {
        let type_name = self.advance().value;
        let name = self.expect(TokenType::Ident).value;
        self.expect(TokenType::Eq);
        let val = self.parse_expr();
        self.match_token(TokenType::Semicolon);

        let mut btype = BType::Any;
        let mut is_bel = false;
        let mut is_ster = false;
        match type_name.as_str() {
            "float" => btype = BType::Float,
            "string" => btype = BType::String,
            "bool" => btype = BType::Bool,
            "bel" => {
                btype = BType::Bel;
                is_bel = true;
            }
            "ster" => {
                btype = BType::Ster;
                is_ster = true;
            }
            "any" => btype = BType::Any,
            _ => {}
        }
        ASTNode::VarDecl(Box::new(VarDecl {
            name,
            var_type: btype,
            value: Box::new(val),
            has_type: true,
            is_bel,
            is_ster,
        }))
    }

    fn parse_fn_decl(&mut self) -> ASTNode {
        self.advance();
        let name = self.expect(TokenType::Ident).value;
        self.expect(TokenType::LParen);

        let mut params = Vec::new();
        if self.peek().tt != TokenType::RParen {
            let p_name = self.expect(TokenType::Ident).value;
            let mut p_type = BType::Any;
            if self.match_token(TokenType::Colon) {
                p_type = self.parse_type();
            }
            params.push(TypedParam {
                name: p_name,
                param_type: p_type,
            });
            while self.match_token(TokenType::Comma) {
                let p_name = self.expect(TokenType::Ident).value;
                let mut p_type = BType::Any;
                if self.match_token(TokenType::Colon) {
                    p_type = self.parse_type();
                }
                params.push(TypedParam {
                    name: p_name,
                    param_type: p_type,
                });
            }
        }
        self.expect(TokenType::RParen);

        let mut ret_type = BType::Any;
        let mut has_ret_type = false;
        if self.match_token(TokenType::Colon) {
            ret_type = self.parse_type();
            has_ret_type = true;
        }

        let body = self.parse_block();
        ASTNode::FnDecl(Box::new(FnDecl {
            name,
            params,
            return_type: ret_type,
            has_ret_type,
            body: Box::new(body),
        }))
    }

    fn parse_return(&mut self) -> ASTNode {
        self.advance();
        let mut val = None;
        if self.peek().tt != TokenType::Semicolon
            && self.peek().tt != TokenType::RBrace
            && self.peek().tt != TokenType::Eof
        {
            val = Some(Box::new(self.parse_expr()));
        }
        self.match_token(TokenType::Semicolon);
        ASTNode::ReturnStmt(Box::new(ReturnStmt { value: val }))
    }

    fn parse_expr(&mut self) -> ASTNode {
        self.parse_or()
    }

    fn parse_or(&mut self) -> ASTNode {
        let mut left = self.parse_and();
        while self.peek().tt == TokenType::Or {
            self.advance();
            let right = self.parse_and();
            left = ASTNode::BinExpr(Box::new(BinExpr {
                op: "||".to_string(),
                left: Box::new(left),
                right: Box::new(right),
            }));
        }
        left
    }

    fn parse_and(&mut self) -> ASTNode {
        let mut left = self.parse_comparison();
        while self.peek().tt == TokenType::And {
            self.advance();
            let right = self.parse_comparison();
            left = ASTNode::BinExpr(Box::new(BinExpr {
                op: "&&".to_string(),
                left: Box::new(left),
                right: Box::new(right),
            }));
        }
        left
    }

    fn parse_comparison(&mut self) -> ASTNode {
        let mut left = self.parse_add_sub();
        loop {
            let tt = self.peek().tt;
            if tt == TokenType::EqEq
                || tt == TokenType::Neq
                || tt == TokenType::Lt
                || tt == TokenType::Gt
                || tt == TokenType::Lte
                || tt == TokenType::Gte
            {
                let op = self.advance().value;
                let right = self.parse_add_sub();
                left = ASTNode::BinExpr(Box::new(BinExpr {
                    op,
                    left: Box::new(left),
                    right: Box::new(right),
                }));
            } else {
                break;
            }
        }
        left
    }

    fn parse_add_sub(&mut self) -> ASTNode {
        let mut left = self.parse_mul_div();
        loop {
            let tt = self.peek().tt;
            if tt == TokenType::Plus || tt == TokenType::Minus {
                let op = self.advance().value;
                let right = self.parse_mul_div();
                left = ASTNode::BinExpr(Box::new(BinExpr {
                    op,
                    left: Box::new(left),
                    right: Box::new(right),
                }));
            } else {
                break;
            }
        }
        left
    }

    fn parse_mul_div(&mut self) -> ASTNode {
        let mut left = self.parse_unary();
        loop {
            let tt = self.peek().tt;
            if tt == TokenType::Star || tt == TokenType::Slash || tt == TokenType::Percent {
                let op = self.advance().value;
                let right = self.parse_unary();
                left = ASTNode::BinExpr(Box::new(BinExpr {
                    op,
                    left: Box::new(left),
                    right: Box::new(right),
                }));
            } else {
                break;
            }
        }
        left
    }

    fn parse_unary(&mut self) -> ASTNode {
        if self.peek().tt == TokenType::Minus {
            self.advance();
            return ASTNode::UnaryExpr(Box::new(UnaryExpr {
                op: "-".to_string(),
                expr: Box::new(self.parse_postfix()),
            }));
        }
        if self.peek().tt == TokenType::Not {
            self.advance();
            return ASTNode::UnaryExpr(Box::new(UnaryExpr {
                op: "!".to_string(),
                expr: Box::new(self.parse_postfix()),
            }));
        }
        self.parse_postfix()
    }

    fn parse_postfix(&mut self) -> ASTNode {
        let mut node = self.parse_primary();
        loop {
            if self.peek().tt == TokenType::LParen {
                self.advance();
                let name = match &node {
                    ASTNode::Ident(i) => i.name.clone(),
                    ASTNode::CallExpr(c) => c.name.clone(),
                    _ => {
                        panic!("line {}: not callable", self.peek().line);
                    }
                };
                let mut args = Vec::new();
                if self.peek().tt != TokenType::RParen {
                    args.push(self.parse_expr());
                    while self.match_token(TokenType::Comma) {
                        args.push(self.parse_expr());
                    }
                }
                self.expect(TokenType::RParen);
                node = ASTNode::CallExpr(Box::new(CallExpr { name, args }));
            } else if self.peek().tt == TokenType::Dot {
                self.advance();
                let right = self.expect(TokenType::Ident).value;
                node = ASTNode::DotExpr(Box::new(DotExpr {
                    left: Box::new(node),
                    right,
                }));
            } else if self.peek().tt == TokenType::LBracket {
                self.advance();
                let idx = self.parse_expr();
                self.expect(TokenType::RBracket);
                node = ASTNode::IndexExpr(Box::new(IndexExpr {
                    obj: Box::new(node),
                    index: Box::new(idx),
                }));
            } else {
                break;
            }
        }
        node
    }

    fn parse_primary(&mut self) -> ASTNode {
        let t = self.peek().clone();
        match t.tt {
            TokenType::String => {
                self.advance();
                ASTNode::StringLit(Box::new(StringLit { value: t.value }))
            }
            TokenType::Number => {
                self.advance();
                ASTNode::NumLit(Box::new(NumLit { value: t.value }))
            }
            TokenType::Ident => {
                self.advance();
                match t.value.as_str() {
                    "true" => ASTNode::BoolLit(Box::new(BoolLit { value: true })),
                    "false" => ASTNode::BoolLit(Box::new(BoolLit { value: false })),
                    "null" => ASTNode::Ident(Box::new(Ident {
                        name: "null".to_string(),
                    })),
                    _ => ASTNode::Ident(Box::new(Ident { name: t.value })),
                }
            }
            TokenType::LParen => {
                self.advance();
                let expr = self.parse_expr();
                self.expect(TokenType::RParen);
                expr
            }
            TokenType::LBracket => self.parse_array_lit(),
            _ => {
                panic!(
                    "line {}:{}: unexpected token '{}' ({})",
                    t.line,
                    t.col,
                    t.tt.name(),
                    t.value
                );
            }
        }
    }

    fn parse_array_lit(&mut self) -> ASTNode {
        self.expect(TokenType::LBracket);
        let mut elems = Vec::new();
        if self.peek().tt != TokenType::RBracket {
            elems.push(self.parse_expr());
            while self.match_token(TokenType::Comma) {
                elems.push(self.parse_expr());
            }
        }
        self.expect(TokenType::RBracket);
        ASTNode::ArrayLit(Box::new(ArrayLit { elems }))
    }
}
